### ISSUE TITLE
Replace JSON module with OJ

### ISSUE THREAD

**Issue by azumakuniyuki** [sisimai/rb-Sisimai#42]
For getting JSON string more faster
http://qiita.com/abey1192/items/0475b401d9b7e8d33cbd

**Comment by azumakuniyuki** [comment]
Other codes are slow, maybe.
```json-vs-oj.rb
require 'benchmark'
require 'oj'
require 'json'
require 'sisimai'

HOWMANYTIMES = 1000
TO_BE_PARSED = './set-of-emails/maildir/bsd'

data = Sisimai.make(TO_BE_PARSED)

Benchmark.bmbm do |b|
  b.report('Encode using JSON') do
    HOWMANYTIMES.times do
      j = JSON::State.new
      j.space = ' '
      j.object_nl = ' '
      JSON.generate(data, j)
    end
  end

  b.report('Encode using Oj') do
    HOWMANYTIMES.times do
      Oj.dump(data, :mode => :compat)
    end
  end
end
```
```
Rehearsal -----------------------------------------------------
Encode using JSON  31.600000   0.680000  32.280000 ( 32.311724)
Encode using Oj    31.950000   0.690000  32.640000 ( 32.684706)
------------------------------------------- total: 64.920000sec

                        user     system      total        real
Encode using JSON  33.770000   0.810000  34.580000 ( 34.778954)
Encode using Oj    35.370000   0.850000  36.220000 ( 36.334508)
ruby json-vs-oj.rb  134.00s user 3.24s system 99% cpu 2:17.89 total
```

**Comment by hiroyuki-sato** [comment]
Oj call ``to_hash`` method if it is custom class.
If it fail, then call to_json method.

> :compat mode attempts to be compatible with other systems. It will serialize any Object, but will check to see if the Object implements an to_hash or to_json method. If either exists, that method is used for serializing the Object. Since as_json is more flexible and produces more consistent output, it is preferred over the to_json method. If neither the to_json or to_hash methods exists, then the Oj internal Object variable encoding is used.

[Options](https://github.com/ohler55/oj#options)

This is test code.

```ruby
require 'oj'
require 'pp'

class Test
  attr_accessor :val1, :val2
  def to_json
    pp '**json**'
    '{}'
  end
  def to_hash
    pp '***'
    { 'val1' => val1.iso8601, 'val2' => val2.to_s }
  end
end

test = Test.new
test.val1 = Time.new
test.val2 = 'hoge2'
```

```
ruby test1.rb 
"***"
{"val1":"2016-02-09T22:12:34+09:00","val2":"hoge2"}
```

Maybe current implementation call this part.
https://github.com/sisimai/rb-Sisimai/blob/master/lib/sisimai/data/json.rb#L25

This is called JSON library, Even you use Oj gem.

**Comment by hiroyuki-sato** [comment]
Additonal Japanese note.

> :compat
> compatはcompatibilityの略
> 渡されたObjectがto_hashかto_jsonを呼べればそれを使う。なければOjがよしなにしてくれる

http://d.hatena.ne.jp/rightgo09_ruby/20131110/p1

**Comment by azumakuniyuki** [comment]
@hiroyuki-sato Thanks for your advices !

**Comment by azumakuniyuki** [comment]
Oj is faster than JSON
```json-vs-oj.rb
require 'benchmark'
require 'oj'
require 'json'

HOWMANYTIMES = 200000
neko = { 'neko' => 'kijitora', 'nyaa' => 'meow', 'cats' => 4401 }
nyan = [
  { 'neko' => 'mikeneko', 'nyaa' => 'nyan', 'cats' => 3306 },
  { 'neko' => 'sironeko', 'nyaa' => 'shaa', 'cats' => 4605 },
]


Benchmark.bmbm do |b|
  b.report('Encode using JSON') do
    HOWMANYTIMES.times do
      j = JSON::State.new
      j.space = ' '
      j.object_nl = ' '
      JSON.generate(neko, j)
      JSON.generate(nyan, j)
    end
  end

  b.report('Encode using Oj') do
    HOWMANYTIMES.times do
      Oj.dump(neko, :mode => :compat)
      Oj.dump(nyan, :mode => :compat)
    end
  end
end
```
Rehearsal -----------------------------------------------------
Encode using JSON   3.200000   0.030000   3.230000 (  3.269022)
Encode using Oj     1.330000   0.030000   1.360000 (  1.368235)
-------------------------------------------- total: 4.590000sec

                        user     system      total        real
Encode using JSON   3.040000   0.030000   3.070000 (  3.094750)
Encode using Oj     1.240000   0.020000   1.260000 (  1.266607)
ruby json-vs-oj.rb  8.92s user 0.15s system 98% cpu 9.233 total
```

```

**Comment by azumakuniyuki** [comment]
Sisimai::Data#to_hash is implemented at branch `using-oj` but the results of benchmark are almost the same with code which have no `to_hash` method in Sisimai::Data.
```
require 'benchmark'
require 'oj'
require 'json'
require 'sisimai'

HOWMANYTIMES = 200
TO_BE_PARSED = './set-of-emails/maildir/bsd'
data = Sisimai.make(TO_BE_PARSED)

Benchmark.bmbm do |b|
  b.report('Encode using JSON') do
    HOWMANYTIMES.times do
      j = JSON::State.new
      j.space = ' '
      j.object_nl = ' '
      JSON.generate(data, j)
    end
  end

  b.report('Encode using Oj') do
    HOWMANYTIMES.times do
      Oj.dump(data, :mode => :compat)
    end
  end
end
```
```
Rehearsal -----------------------------------------------------
Encode using JSON   6.260000   0.140000   6.400000 (  6.409495)
Encode using Oj     6.300000   0.140000   6.440000 (  6.438790)
------------------------------------------- total: 12.840000sec

                        user     system      total        real
Encode using JSON   6.250000   0.150000   6.400000 (  6.406366)
Encode using Oj     6.290000   0.120000   6.410000 (  6.411806)
ruby json-vs-oj.rb  26.41s user 0.77s system 99% cpu 27.406 total
```

**Comment by azumakuniyuki** [comment]
I have quit forgot to add the following line at the top of the benchmark script
```
$:.unshift './lib'
```

Oj is certainly faster than JSON.
```
Rehearsal -----------------------------------------------------
Encode using JSON  39.340000   5.060000  44.400000 ( 44.813980)
Encode using Oj    26.440000   0.700000  27.140000 ( 27.268303)
------------------------------------------- total: 71.540000sec

                        user     system      total        real
Encode using JSON  39.000000   5.120000  44.120000 ( 44.442563)
Encode using Oj    26.010000   0.690000  26.700000 ( 26.832384)
ruby json-vs-oj.rb  132.10s user 11.85s system 99% cpu 2:25.18 total
```

**Comment by azumakuniyuki** [comment]
However, Oj does not work on JRuby because
> Oj is compatible with Ruby 1.8.7, 1.9.2, 1.9.3, 2.0.0, 2.1, 2.2, 2.3 and RBX. Support for JRuby has been removed as JRuby no longer supports C extensions and there are bugs in the older versions that are not being fixed.
described at https://github.com/ohler55/oj and build tests have failed at https://travis-ci.org/sisimai/rb-Sisimai/jobs/108080022

**Comment by hiroyuki-sato** [comment]
OMG, I forgot all about it. (Oj does not work on JRuby).
Maybe you can write dependency like the following. 

https://www.ruby-forum.com/topic/4414732#1112249

```ruby
unless /java/ === RUBY_PLATFORM
  spec.add_dependency 'oj'
end
```

And [JrJackson](https://github.com/guyboertje/jrjackson) is popular JSON gem on JRuby platform.
I'll check this.

**Comment by azumakuniyuki** [comment]
@hiroyuki-sato I did not know about JrJackson, thanks for the advice. I will check the library and try to implement code using JrJackson under JRuby environment.

**Comment by azumakuniyuki** [comment]
JrJackson is also fast !
```json-vs-jrjackson.rb
$:.unshift './lib'
require 'benchmark'
require 'json'
require 'jrjackson'
require 'sisimai'

HOWMANYTIMES = 1000
TO_BE_PARSED = './set-of-emails/maildir/bsd'
data = Sisimai.make(TO_BE_PARSED)

Benchmark.bmbm do |b|
  b.report('Encode using JSON') do
    HOWMANYTIMES.times do
      JSON.generate(data)
    end
  end

  b.report('Encode using JrJackson') do
    HOWMANYTIMES.times do
      JrJackson::Json.dump(data)
    end
  end

end
```
```
Rehearsal ----------------------------------------------------------
Encode using JSON       43.000000   0.730000  43.730000 ( 37.583855)
Encode using JrJackson  20.930000   0.290000  21.220000 ( 21.397240)
------------------------------------------------ total: 64.950000sec

                             user     system      total        real
Encode using JSON       37.070000   0.530000  37.600000 ( 37.551781)
Encode using JrJackson  19.110000   0.310000  19.420000 ( 19.477917)
jruby json-vs-jrjackson.rb  167.03s user 3.60s system 127% cpu 2:14.22 total
```

**Comment by hiroyuki-sato** [comment]
👍

**Comment by azumakuniyuki** [comment]
Implemented codes have been marged at Pull-Request #44
